---
hide:
  - navigation
  - toc 
---

# Contact

The complete Imputation Server 2 source code is available on [GitHub](https://github.com/genepi/imputationserver2) and has been developed by [Lukas Forer](https://genepi.i-med.ac.at/team/forer-lukas/) and [Sebastian Schönherr](https://genepi.i-med.ac.at/team/schoenherr-sebastian/) from the Institute of Genetic Epidemiology, Medical University of Innsbruck. 

Feel free to create issues and pull requests. Before contacting us, please have a look at the [FAQ page](faq) first. 

## Michigan Imputation Server Team

Michigan Imputation Server provides a free genotype imputation service using Minimac4. You can upload phased or unphased GWAS genotypes and receive phased and imputed genomes in return. For all uploaded data sets an extensive QC is performed.

* [Christian Fuchsberger](mailto:cfuchsb@umich.edu)
* [Lukas Forer](mailto:lukas.forer@i-med.ac.at)
* [Sebastian Schönherr](mailto:sebastian.schoenherr@i-med.ac.at)
* [Sayantan Das](mailto:sayantan@umich.edu)
* [Gonçalo Abecasis](mailto:goncalo@umich.edu)

Please contact [Christian Fuchsberger](mailto:cfuchsb@umich.edu) in case of other problems.


## TOPMed Imputation Server Team

Michigan Imputation Server provides a free genotype imputation service using Minimac4. You can upload phased or unphased GWAS genotypes and receive phased and imputed genomes in return. For all uploaded data sets an extensive QC is performed.

* [Albert Smith](mailto:albertvs@umich.edu)
* [Andy Boughton](mailto:abought@umich.edu)

Please use [this addrees](mailto:imputationserver@umich.edu) for all inquiries.


## Imputation engine: [Minimac4](http://genome.sph.umich.edu/wiki/Minimac4)

Minimac4 is a lower memory and more computationally efficient implementation of the genotype imputation algorithms in minimac/mininac2/minimac3.

* [Sayantan Das](mailto:sayantan@umich.edu)
* [Christian Fuchsberger](mailto:cfuchsb@umich.edu)
* [Gonçalo Abecasis](mailto:goncalo@umich.edu)

## Cloud framework: [Cloudgene](https://www.cloudgene.io/)

Cloudgene is a framework to build Software As A Service (SaaS) platforms for data analysis pipelines. By connecting command-line programs, scripts or Hadoop applications to Cloudgene, a powerful web application can be created within minutes. Cloudgene supports the complete workflow including data transfer, program execution and data export. Cloudgene is developed at the Division of Genetic Epidemiology Innsbruck in cooperation with the Center for Statistical Genetics, University of Michigan.

* [Lukas Forer](mailto:lukas.forer@i-med.ac.at)
* [Sebastian Schönherr](mailto:sebastian.schoenherr@i-med.ac.at)

## Phasing engine: [Eagle2](https://data.broadinstitute.org/alkesgroup/Eagle/)

For haplotype phasing Eagle2 is used. Eagle2 attains high accuracy across a broad range of cohort sizes by efficiently leveraging information from large external reference panels (such as the Haplotype Reference Consortium; HRC).
